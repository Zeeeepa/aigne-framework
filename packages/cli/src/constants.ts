import { AnthropicChatModel } from "@aigne/anthropic";
import { BedrockChatModel } from "@aigne/bedrock";
import { DeepSeekChatModel } from "@aigne/deepseek";
import { GeminiChatModel } from "@aigne/gemini";
import { OllamaChatModel } from "@aigne/ollama";
import { OpenRouterChatModel } from "@aigne/open-router";
import { OpenAIChatModel } from "@aigne/openai";
import { XAIChatModel } from "@aigne/xai";
import pkg from "../package.json" with { type: "json" };

export const AIGNE_CLI_VERSION = pkg.version;

export const availableModels = [
  OpenAIChatModel,
  AnthropicChatModel,
  BedrockChatModel,
  DeepSeekChatModel,
  GeminiChatModel,
  OllamaChatModel,
  OpenRouterChatModel,
  XAIChatModel,
];
