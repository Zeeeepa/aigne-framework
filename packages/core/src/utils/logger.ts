import { nodejs } from "@aigne/platform-helpers/nodejs/index.js";
import debug from "debug";

export enum LogLevel {
  ERROR = "error",
  WARN = "warn",
  INFO = "info",
  DEBUG = "debug",
}

const levels = Object.values(LogLevel);

export function getLevelFromEnv(ns: string): LogLevel | undefined {
  for (const level of levels.toReversed()) {
    if (debug.enabled(`${ns}:${level}`)) {
      return level;
    }
  }
}

export class Logger {
  constructor(public options: { level: LogLevel; ns: string }) {
    this.debugLogger = debug(`${options.ns}:${LogLevel.DEBUG}`);
    this.infoLogger = debug(`${options.ns}:${LogLevel.INFO}`);
    this.warnLogger = debug(`${options.ns}:${LogLevel.WARN}`);
    this.errorLogger = debug(`${options.ns}:${LogLevel.ERROR}`);

    this.level = getLevelFromEnv(options.ns) || options.level;

    for (const logger of [this.debugLogger, this.infoLogger, this.warnLogger]) {
      // @ts-ignore
      logger.useColors = nodejs.isStdoutATTY;
      logger.enabled = true;
      logger.log = (...args: unknown[]) => this.logMessage(...args);
    }

    this.errorLogger.log = (...args: unknown[]) => this.logError(...args);
    // @ts-ignore
    this.errorLogger.useColors = nodejs.isStderrATTY;
    this.errorLogger.enabled = true;
  }

  level: LogLevel;

  private debugLogger: debug.Debugger;

  private infoLogger: debug.Debugger;

  private warnLogger: debug.Debugger;

  private errorLogger: debug.Debugger;

  logMessage = console.log;

  logError = console.error;

  debug(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.DEBUG)) {
      this.debugLogger(message, ...args);
    }
  }

  info(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.INFO)) {
      this.infoLogger(message, ...args);
    }
  }

  warn(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.WARN)) {
      this.warnLogger(message, ...args);
    }
  }

  error(message: string, ...args: unknown[]) {
    if (this.enabled(LogLevel.ERROR)) {
      this.errorLogger(message, ...args);
    }
  }

  enabled(level: LogLevel) {
    return levels.indexOf(this.level) >= levels.indexOf(level);
  }
}

export const logger = new Logger({
  ns: "aigne:core",
  level: LogLevel.INFO,
});
