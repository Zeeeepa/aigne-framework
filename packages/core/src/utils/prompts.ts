export function getJsonOutputPrompt(schema: Record<string, unknown> | string) {
  let prompt = "Output must be a JSON object containing the following fields only.";
  if (typeof schema === "string") {
    prompt += `\n<json_fields>\n${schema}\n</json_fields>`;
  } else {
    prompt += `\n<json_fields>\n${JSON.stringify(schema)}\n</json_fields>`;
  }
  return prompt;
}
