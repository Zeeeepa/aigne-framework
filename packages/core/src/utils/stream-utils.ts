import equal from "fast-deep-equal";
import {
  type AgentProcessAsyncGenerator,
  type AgentResponseChunk,
  type AgentResponseStream,
  isAgentResponseDelta,
  isEmptyChunk,
  type Message,
} from "../agents/agent.js";
import { isNonNullable, isRecord, omitBy, type PromiseOrValue } from "./type-utils.js";
import "./stream-polyfill.js";
import type { ReadableStreamDefaultReadResult } from "bun";

export function objectToAgentResponseStream<T extends Message>(json: T): AgentResponseStream<T> {
  if (!isRecord(json)) {
    throw new Error(
      `expect to return a record type such as {result: ...}, but got (${typeof json}): ${json}`,
    );
  }

  return new ReadableStream({
    pull(controller) {
      controller.enqueue({ delta: { json } });
      controller.close();
    },
  });
}

export function mergeAgentResponseChunk<T extends Message>(
  output: T,
  chunk: AgentResponseChunk<T>,
) {
  if (isAgentResponseDelta(chunk)) {
    if (chunk.delta.text) {
      for (const [key, text] of Object.entries(chunk.delta.text)) {
        const original = output[key] as string | undefined;
        const t = (original || "") + (text || "");
        if (t) Object.assign(output, { [key]: t });
      }
    }

    if (chunk.delta.json) {
      Object.assign(
        output,
        omitBy(chunk.delta.json, (v) => v === undefined),
      );
    }
  }

  return output;
}

export async function agentResponseStreamToObject<T extends Message>(
  stream: AgentResponseStream<T> | AgentProcessAsyncGenerator<T>,
): Promise<T> {
  const json: T = {} as T;

  if (stream instanceof ReadableStream) {
    for await (const value of stream) {
      mergeAgentResponseChunk(json, value);
    }
  } else {
    for (;;) {
      const chunk = await stream.next();
      if (chunk.value) {
        if (chunk.done) {
          Object.assign(json, chunk.value);
        } else {
          mergeAgentResponseChunk(json, chunk.value);
        }
      }
      if (chunk.done) break;
    }
  }

  return json;
}

export function asyncGeneratorToReadableStream<T extends Message>(
  generator: AgentProcessAsyncGenerator<T>,
): AgentResponseStream<T> {
  return new ReadableStream({
    async pull(controller) {
      try {
        const chunk = await generator.next();
        if (chunk.value) {
          if (chunk.done) {
            controller.enqueue({ delta: { json: chunk.value } });
          } else {
            controller.enqueue(chunk.value);
          }
        }

        if (chunk.done) {
          controller.close();
        }
      } catch (error) {
        controller.error(error);
      }
    },
  });
}

export function onAgentResponseStreamEnd<T extends Message>(
  stream: AgentResponseStream<T>,
  options?: {
    onChunk?: (
      chunk: AgentResponseChunk<T>,
    ) => PromiseOrValue<AgentResponseChunk<T> | undefined | void>;
    onResult?: (result: T) => PromiseOrValue<Partial<T> | undefined | void>;
    onError?: (error: Error) => PromiseOrValue<Error>;
  },
): AgentResponseStream<T> {
  const json: T = {} as T;
  const reader = stream.getReader();

  return new ReadableStream({
    async pull(controller) {
      try {
        while (true) {
          const { value, done } = await reader.read();

          if (done) {
            const result = (await options?.onResult?.(json)) ?? json;

            if (result && !equal(result, json)) {
              let chunk: AgentResponseChunk<T> = { delta: { json: result } };
              chunk = (await options?.onChunk?.(chunk)) ?? chunk;
              controller.enqueue(chunk);
            }

            controller.close();

            return;
          }

          mergeAgentResponseChunk(json, value);

          const chunk = (await options?.onChunk?.(value)) ?? value;
          if (!isEmptyChunk(chunk)) {
            controller.enqueue(chunk);
            break;
          }
        }
      } catch (error) {
        controller.error((await options?.onError?.(error)) ?? error);
      }
    },
  });
}

export function isAsyncGenerator<T extends AsyncGenerator>(
  value: AsyncGenerator | unknown,
): value is T {
  return typeof value === "object" && value !== null && Symbol.asyncIterator in value;
}

export async function* arrayToAgentProcessAsyncGenerator<T extends Message>(
  chunks: (AgentResponseChunk<T> | Error)[],
  result?: Partial<T>,
): AgentProcessAsyncGenerator<T> {
  for (const chunk of chunks) {
    if (chunk instanceof Error) throw chunk;

    yield chunk;
  }
  if (result !== undefined) return result;
}

export function arrayToReadableStream<T>(chunks: (T | Error)[]): ReadableStream<T> {
  const list = [...chunks];

  return new ReadableStream({
    pull(controller) {
      const item = list.shift();
      if (!item) {
        controller.close();
        return;
      }

      if (item instanceof Error) {
        controller.error(item);
        return;
      }

      controller.enqueue(item);
    },
  });
}

export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options: { catchError: true },
): Promise<(T | Error)[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<T[]>;
export async function readableStreamToArray<T>(
  stream: ReadableStream<T>,
  options?: { catchError?: boolean },
): Promise<(T | Error)[]> {
  const result: T[] = [];
  try {
    for await (const value of stream) {
      result.push(value);
    }
  } catch (error) {
    if (!options?.catchError) throw error;

    result.push(error);
  }
  return result;
}

export function stringToAgentResponseStream(
  str: string,
  key: "text" | string = "text",
): AgentResponseStream<Message> {
  const segmenter = new Intl.Segmenter(undefined, { granularity: "word" });
  const segments = segmenter.segment(str);

  return arrayToReadableStream(
    Array.from(segments).map((segment) => ({ delta: { text: { [key]: segment.segment } } })),
  );
}

export function toReadableStream(stream: NodeJS.ReadStream) {
  return new ReadableStream<Uint8Array>({
    start(controller) {
      const onData = (chunk: Buffer) => {
        controller.enqueue(new Uint8Array(chunk));
      };

      const onEnd = () => {
        cleanup();
        controller.close();
      };

      const onError = (err: Error) => {
        cleanup();
        controller.error(err);
      };

      function cleanup() {
        stream.off("data", onData);
        stream.off("end", onEnd);
        stream.off("error", onError);
      }

      stream.on("data", onData);
      stream.on("end", onEnd);
      stream.on("error", onError);
    },
  });
}

export async function readAllString(stream: NodeJS.ReadStream | ReadableStream): Promise<string> {
  return (
    await readableStreamToArray(
      (stream instanceof ReadableStream ? stream : toReadableStream(stream)).pipeThrough(
        new TextDecoderStream(),
      ),
    )
  ).join("");
}

export function mergeReadableStreams<T1, T2>(
  s1: ReadableStream<T1>,
  s2: ReadableStream<T2>,
): ReadableStream<T1 | T2>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any>;
export function mergeReadableStreams(
  ...streams: (ReadableStream<any> | undefined)[]
): ReadableStream<any> {
  type Reader = {
    reader: ReadableStreamDefaultReader;
    reading?: Promise<{
      result: ReadableStreamDefaultReadResult<any>;
      item: any;
    }>;
  };

  let readers: Reader[] | undefined;

  return new ReadableStream({
    async pull(controller) {
      try {
        readers ??= streams.filter(isNonNullable).map((s) => ({ reader: s.getReader(), data: [] }));

        while (readers.length) {
          const chunk: Awaited<NonNullable<Reader["reading"]>> = await Promise.race(
            readers.map((i) => {
              i.reading ??= i.reader.read().then((result) => ({ result, item: i }));
              return i.reading;
            }),
          );

          if (chunk.result.value) {
            controller.enqueue(chunk.result.value);
            chunk.item.reading = undefined;
            return;
          }

          if (chunk.result.done) {
            readers = readers.filter((i) => i !== chunk.item);
          }
        }

        controller.close();
      } catch (error) {
        controller.error(error);
        if (readers) for (const item of readers) item.reader.releaseLock();
      }
    },
  });
}
