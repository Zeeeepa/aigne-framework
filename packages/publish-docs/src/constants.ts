// Component mount point ID for Discuss Kit
export const DISCUSS_KIT_DID = "z8ia1WEiBZ7hxURf6LwH21Wpg99vophFwSJdu";

// Component DID for Media Kit
export const MEDIA_KIT_DID = "z8ia1mAXo8ZE7ytGF36L5uBf9kD2kenhqFGp9";
