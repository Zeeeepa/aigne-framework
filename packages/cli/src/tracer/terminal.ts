import { EOL } from "node:os";
import { type InspectOptions, inspect } from "node:util";
import {
  type Agent,
  type AgentHooks,
  AIAgent,
  ChatModel,
  type ChatModelOutput,
  type Context,
  type ContextUsage,
  DEFAULT_OUTPUT_KEY,
  type InvokeOptions,
  type Message,
  mergeContextUsage,
  newEmptyContextUsage,
  UserAgent,
} from "@aigne/core";
import { promiseWithResolvers } from "@aigne/core/utils/promise.js";
import { flat, omit } from "@aigne/core/utils/type-utils.js";
import { figures, type Listr } from "@aigne/listr2";
import { markedTerminal } from "@aigne/marked-terminal";
import * as prompts from "@inquirer/prompts";
import chalk from "chalk";
import { Marked } from "marked";
import { AIGNE_HUB_CREDITS_NOT_ENOUGH_ERROR_TYPE } from "../constants.js";
import { AIGNEListr, AIGNEListrRenderer, type AIGNEListrTaskWrapper } from "../utils/listr.js";
import { highlightUrl } from "../utils/string-utils.js";
import { parseDuration } from "../utils/time.js";

const CREDITS_ERROR_PROCESSED_FLAG = "$credits_error_processed";

export interface TerminalTracerOptions {
  outputKey?: string;
}

export class TerminalTracer {
  constructor(
    public readonly context: Context,
    public readonly options: TerminalTracerOptions = {},
  ) {}

  private tasks: { [callId: string]: Task } = {};

  async run(agent: Agent, input: Message, options?: InvokeOptions) {
    await this.context.observer?.serve();

    const context = this.context.newContext({ reset: true });

    const listr = new AIGNEListr(
      {
        formatRequest: (options?: { running?: boolean }) =>
          this.formatRequest(agent, context, input, options),
        formatResult: (result, options?: { running?: boolean }) =>
          [this.formatResult(agent, context, result, options)].filter(Boolean),
      },
      [],
      { concurrent: true, exitOnError: false },
    );
    this.listr = listr;

    const collapsedMap = new Map<
      string,
      { ancestor: { contextId: string }; usage: ContextUsage; models: Set<string> }
    >();
    const hideContextIds = new Set<string>();

    const onStart: AgentHooks["onStart"] = async ({ context, agent, ...event }) => {
      if (agent instanceof UserAgent) return;

      if (agent.taskRenderMode === "hide") {
        hideContextIds.add(context.id);
        return;
      } else if (agent.taskRenderMode === "collapse") {
        collapsedMap.set(context.id, {
          ancestor: { contextId: context.id },
          usage: newEmptyContextUsage(),
          models: new Set(),
        });
      }

      if (context.parentId) {
        if (hideContextIds.has(context.parentId)) {
          hideContextIds.add(context.id);
          return;
        }

        const collapsed = collapsedMap.get(context.parentId);
        if (collapsed) {
          collapsedMap.set(context.id, collapsed);
          return;
        }
      }

      const contextId = context.id;
      const parentContextId = context.parentId;

      const task: Task = {
        ...promiseWithResolvers(),
        agent,
        input: event.input,
        listr: promiseWithResolvers(),
        startTime: Date.now(),
      };
      this.tasks[contextId] = task;

      const listrTask: Parameters<typeof listr.add>[0] = {
        title: await this.formatTaskTitle(agent, { ...event }),
        task: (ctx, taskWrapper) => {
          const subtask = taskWrapper.newListr([{ task: () => task.promise }]);
          task.listr.resolve({ subtask, taskWrapper, ctx });
          return subtask;
        },
        rendererOptions: {
          persistentOutput: true,
          outputBar: Number.POSITIVE_INFINITY,
          bottomBar: Number.POSITIVE_INFINITY,
        },
      };

      const parentTask = parentContextId ? this.tasks[parentContextId] : undefined;
      if (parentTask) {
        parentTask.listr.promise.then(({ subtask }) => {
          subtask.add(listrTask);
        });
      } else {
        listr.add(listrTask);
      }

      return { options: { prompts: this.proxiedPrompts } };
    };

    const onSuccess: AgentHooks["onSuccess"] = async ({ context, agent, output, ...event }) => {
      const contextId = context.id;
      const parentContextId = context.parentId;

      const collapsed = collapsedMap.get(contextId);
      if (collapsed) {
        if (agent instanceof ChatModel) {
          const { usage, model } = output as ChatModelOutput;
          if (usage) mergeContextUsage(collapsed.usage, usage);
          if (model) collapsed.models.add(model);
        }

        const task = this.tasks[collapsed.ancestor.contextId];
        if (task) {
          task.usage = collapsed.usage;
          task.extraTitleMetadata ??= {};
          if (collapsed.models.size)
            task.extraTitleMetadata.model = [...collapsed.models].join(",");

          const { taskWrapper } = await task.listr.promise;

          taskWrapper.title = await this.formatTaskTitle(task.agent, {
            input: task.input,
            task,
            usage: Boolean(
              task.usage.inputTokens || task.usage.outputTokens || task.usage.aigneHubCredits,
            ),
            time: context.id === collapsed.ancestor.contextId,
          });

          if (context.id === collapsed.ancestor.contextId) {
            task?.resolve();
          }
          return;
        }
      }

      const task = this.tasks[contextId];
      if (!task) return;

      task.endTime = Date.now();

      const { taskWrapper, ctx } = await task.listr.promise;

      if (agent instanceof ChatModel) {
        const { usage, model } = output as ChatModelOutput;
        task.usage = usage;
        task.extraTitleMetadata ??= {};
        if (model) task.extraTitleMetadata.model = model;
      }

      taskWrapper.title = await this.formatTaskTitle(agent, {
        ...event,
        task,
        usage: true,
        time: true,
      });

      if (!parentContextId || !this.tasks[parentContextId]) {
        Object.assign(ctx, output);
      }

      task.resolve();
    };

    const onError: AgentHooks["onError"] = async ({ context, agent, error, ...event }) => {
      if ("type" in error && error.type === AIGNE_HUB_CREDITS_NOT_ENOUGH_ERROR_TYPE) {
        if (!Object.hasOwn(error, CREDITS_ERROR_PROCESSED_FLAG)) {
          Object.defineProperty(error, CREDITS_ERROR_PROCESSED_FLAG, {
            value: true,
            enumerable: false,
          });

          const retry = await this.promptBuyCredits(error);

          console.log("");

          if (retry === "retry") {
            return { retry: true };
          }
        }
      }

      const contextId = context.id;

      const task = this.tasks[contextId];
      if (!task) return;

      task.endTime = Date.now();

      const { taskWrapper } = await task.listr.promise;
      taskWrapper.title = await this.formatTaskTitle(agent, {
        ...event,
        task,
        usage: true,
        time: true,
      });

      task.reject(error);
    };

    const result = await listr.run(() =>
      context.invoke(agent, input, {
        ...options,
        hooks: flat(
          {
            priority: "high",
            onStart,
            onSuccess,
            onError,
          },
          options?.hooks,
        ),
        streaming: true,
        newContext: false,
      }),
    );

    return { result, context };
  }

  private listr?: AIGNEListr;

  private proxiedPrompts = new Proxy(
    {},
    {
      get: (_target, prop) => {
        // biome-ignore lint/performance/noDynamicNamespaceImportAccess: we need to access prompts dynamically
        const method = prompts[prop as keyof typeof prompts] as (...args: any[]) => any;
        if (typeof method !== "function") return undefined;

        return async (config: any) => {
          const renderer =
            this.listr?.["renderer"] instanceof AIGNEListrRenderer
              ? this.listr["renderer"]
              : undefined;
          await renderer?.pause();

          try {
            return await method({ ...config });
          } finally {
            await renderer?.resume();
          }
        };
      },
    },
  ) as typeof prompts;

  private buyCreditsPromptPromise: Promise<"retry" | "exit"> | undefined;

  private async promptBuyCredits(error: Error) {
    // Avoid multiple agents asking for credits, we will only show the prompt once
    this.buyCreditsPromptPromise ??= (async () => {
      const retry = await this.proxiedPrompts.select({
        message: highlightUrl(error.message),
        choices: [
          {
            name: "I have bought some credits, try again",
            value: "retry" as const,
          },
          {
            name: "Exit",
            value: "exit" as const,
          },
        ],
      });

      return retry;
    })();

    return this.buyCreditsPromptPromise.finally(() => {
      // Clear the promise so that we can show the prompt again if needed
      this.buyCreditsPromptPromise = undefined;
    });
  }

  formatTokenUsage(usage: Partial<ContextUsage>, extra?: { [key: string]: string }) {
    const items = [
      [chalk.yellow(usage.inputTokens), chalk.grey("input tokens")],
      [chalk.cyan(usage.outputTokens), chalk.grey("output tokens")],
      usage.aigneHubCredits
        ? [chalk.blue(usage.aigneHubCredits.toFixed()), chalk.grey("AIGNE Hub credits")]
        : undefined,
      usage.agentCalls ? [chalk.magenta(usage.agentCalls), chalk.grey("agent calls")] : undefined,
    ];

    const content = items.filter((i) => !!i).map((i) => i.join(" "));

    if (extra) {
      content.unshift(
        ...Object.entries(extra)
          .filter(([k, v]) => k && v)
          .map(([k, v]) => `${chalk.grey(k)}: ${v}`),
      );
    }

    return `${chalk.grey("(")}${content.join(chalk.green(", "))}${chalk.grey(")")}`;
  }

  formatTimeUsage(startTime: number, endTime: number) {
    const duration = endTime - startTime;
    return chalk.grey(`[${parseDuration(duration)}]`);
  }

  async formatTaskTitle(
    agent: Agent,
    { task, usage, time, input }: { task?: Task; usage?: boolean; time?: boolean; input: Message },
  ) {
    let title = agent.name;

    if (agent.taskTitle) {
      title += ` ${chalk.cyan(await agent.renderTaskTitle(input))}`;
    }

    if (usage && task?.usage)
      title += ` ${this.formatTokenUsage(task.usage, task.extraTitleMetadata)}`;
    if (time && task?.startTime && task.endTime)
      title += ` ${this.formatTimeUsage(task.startTime, task.endTime)}`;

    return title;
  }

  private marked = new Marked().use(
    {
      // marked-terminal does not support code block meta, so we need to strip it
      walkTokens: (token) => {
        if (token.type === "code") {
          if (typeof token.lang === "string") {
            token.lang = token.lang.trim().split(/\s+/)[0];
          }
        }
      },
    },
    markedTerminal(
      { forceHyperLink: false },
      {
        theme: {
          string: chalk.green,
        },
      },
    ),
  );

  get outputKey() {
    return this.options.outputKey || DEFAULT_OUTPUT_KEY;
  }

  formatRequest(agent: Agent, _context: Context, m: Message = {}, { running = false } = {}) {
    const prefix = `${chalk.grey(figures.pointer)} 💬 `;

    const inputKey = agent instanceof AIAgent ? agent.inputKey : undefined;

    const msg = inputKey ? m[inputKey] : undefined;
    const message = inputKey ? omit(m, inputKey) : m;

    const text =
      msg && typeof msg === "string" ? this.marked.parse(msg, { async: false }).trim() : undefined;

    const json =
      Object.keys(message).length > 0
        ? inspect(message, { colors: true, ...(running ? this.runningInspectOptions : undefined) })
        : undefined;

    const r = [text, json].filter(Boolean).join(EOL).trim();
    if (!r) return undefined;

    return `${prefix}${r}`;
  }

  formatResult(agent: Agent, context: Context, m: Message = {}, { running = false } = {}) {
    const { isTTY } = process.stdout;
    const outputKey = this.outputKey || (agent instanceof AIAgent ? agent.outputKey : undefined);

    const prefix = `${chalk.grey(figures.tick)} 🤖 ${this.formatTokenUsage(context.usage)}`;

    const msg = outputKey ? m[outputKey] : undefined;
    const message = outputKey ? omit(m, outputKey) : m;

    const text =
      msg && typeof msg === "string"
        ? isTTY
          ? this.marked.parse(msg, { async: false }).trim()
          : msg
        : undefined;

    const json =
      Object.keys(message).length > 0
        ? inspect(message, { colors: isTTY, ...(running ? this.runningInspectOptions : undefined) })
        : undefined;

    return [prefix, text, json].filter(Boolean).join(EOL.repeat(2));
  }

  protected runningInspectOptions: InspectOptions = {
    maxArrayLength: 3,
    maxStringLength: 200,
  };
}

type Task = ReturnType<typeof promiseWithResolvers<void>> & {
  listr: ReturnType<
    typeof promiseWithResolvers<{
      ctx: object;
      subtask: Listr;
      taskWrapper: AIGNEListrTaskWrapper;
    }>
  >;
  agent: Agent;
  input: Message;
  startTime?: number;
  endTime?: number;
  usage?: Partial<ContextUsage>;
  extraTitleMetadata?: { [key: string]: string };
};
