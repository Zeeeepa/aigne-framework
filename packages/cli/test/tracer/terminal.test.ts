import { expect, spyOn, test } from "bun:test";
import { AIGNE_HUB_CREDITS_NOT_ENOUGH_ERROR_TYPE } from "@aigne/cli/constants.js";
import { TerminalTracer } from "@aigne/cli/tracer/terminal.js";
import { AIAgent, AIGNE, FunctionAgent } from "@aigne/core";
import { LogLevel, logger } from "@aigne/core/utils/logger.js";
import { arrayToAgentProcessAsyncGenerator } from "@aigne/core/utils/stream-utils.js";
import { OpenAIChatModel } from "@aigne/openai";
import * as prompts from "@inquirer/prompts";

test("TerminalTracer should work correctly", async () => {
  const model = new OpenAIChatModel({});

  const aigne = new AIGNE({ model });
  const context = aigne.newContext();

  const testAgent = AIAgent.from({
    inputKey: "message",
  });

  spyOn(model, "process").mockReturnValue(
    Promise.resolve({ text: "hello, this is a test response message" }),
  );

  const userAgent = aigne.invoke(testAgent);

  const tracer = new TerminalTracer(context);

  const { result } = await tracer.run(userAgent, { message: "hello" });

  expect(result).toMatchSnapshot();
});

test("TerminalTracer should raise error correctly", async () => {
  logger.level = LogLevel.INFO;
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const testAgent = AIAgent.from({
    inputKey: "message",
  });

  spyOn(testAgent, "process").mockReturnValueOnce(
    arrayToAgentProcessAsyncGenerator([new Error("test error")]),
  );

  const userAgent = aigne.invoke(testAgent);

  const tracer = new TerminalTracer(context);

  const result = tracer.run(userAgent, { message: "hello" });

  expect(result).rejects.toThrowError("test error");
});

test("TerminalTracer should render output message with markdown highlight", async () => {
  const model = new OpenAIChatModel({});

  const aigne = new AIGNE({ model });
  const context = aigne.newContext();

  const tracer = new TerminalTracer(context);

  const originalIsTTY = process.stdout.isTTY;
  process.stdout.isTTY = true;

  expect(
    tracer.formatResult(new AIAgent({ inputKey: "message" }), context, {
      message: "## Hello\nI am from [**AIGNE**](https://www.aigne.io)",
    }),
  ).toMatchSnapshot();

  process.stdout.isTTY = originalIsTTY;
});

test("TerminalTracer should render output message without markdown highlight in non-tty", async () => {
  logger.level = LogLevel.INFO;
  const model = new OpenAIChatModel({});

  const aigne = new AIGNE({ model });
  const context = aigne.newContext();

  const tracer = new TerminalTracer(context);

  const originalIsTTY = process.stdout.isTTY;
  process.stdout.isTTY = false;

  expect(
    tracer.formatResult(new AIAgent({ outputKey: "message" }), context, {
      message: "## Hello\nI am from [**AIGNE**](https://www.aigne.io)",
    }),
  ).toMatchSnapshot();

  process.stdout.isTTY = originalIsTTY;
});

test("TerminalTracer.marked should stripe code block meta", async () => {
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const tracer = new TerminalTracer(context);

  expect(
    tracer["marked"].parse(`\
hello

${"```"}ts file="test.ts" region="test-region"
function test() {
}
${"```"}
`),
  ).toMatchSnapshot();
});

test("TerminalTracer should", async () => {
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const agent = FunctionAgent.from(async (_, options) => {
    const name = await options.prompts?.input({ message: "What is your name?" });
    const age = await options.prompts?.number({ message: "What is your age?" });
    const color = await options.prompts?.select({
      message: "What is your favorite color?",
      choices: ["red", "green", "blue"],
    });

    return { name, age, color };
  });

  const tracer = new TerminalTracer(context);

  const input = spyOn(prompts, "input").mockReturnValueOnce(Promise.resolve("John Doe") as any);
  const number = spyOn(prompts, "number").mockReturnValueOnce(Promise.resolve(18) as any);
  const select = spyOn(prompts, "select").mockReturnValueOnce(Promise.resolve("red") as any);

  const { result } = await tracer.run(agent, {});

  expect(result).toEqual({
    name: "John Doe",
    age: 18,
    color: "red",
  });

  input.mockRestore();
  number.mockRestore();
  select.mockRestore();
});

test("TerminalTracer should handle buy credits prompt", async () => {
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const agent = FunctionAgent.from(() => ({}));

  const tracer = new TerminalTracer(context);

  const promptBuyCredits = spyOn(tracer, "promptBuyCredits").mockResolvedValueOnce("retry");

  spyOn(agent, "process")
    .mockImplementationOnce(() => {
      const e = new Error("You need to buy credits to continue.");
      (e as any).type = AIGNE_HUB_CREDITS_NOT_ENOUGH_ERROR_TYPE;
      throw e;
    })
    .mockReturnValueOnce({ text: "This is a response after buying credits" });

  const result = await tracer.run(agent, {});

  expect(result.result).toEqual({ text: "This is a response after buying credits" });
  expect(promptBuyCredits).toHaveBeenCalledTimes(1);
});

test("TerminalTracer should exit if user choose exit for buy credits", async () => {
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const agent = FunctionAgent.from(() => ({}));

  const tracer = new TerminalTracer(context);

  const promptBuyCredits = spyOn(tracer, "promptBuyCredits")
    .mockResolvedValueOnce("retry")
    .mockResolvedValueOnce("exit");

  const process = spyOn(agent, "process")
    .mockImplementationOnce(() => {
      const e = new Error("You need to buy credits to continue.");
      (e as any).type = AIGNE_HUB_CREDITS_NOT_ENOUGH_ERROR_TYPE;
      throw e;
    })
    .mockImplementationOnce(() => {
      const e = new Error("You need to buy credits to continue.");
      (e as any).type = AIGNE_HUB_CREDITS_NOT_ENOUGH_ERROR_TYPE;
      throw e;
    });

  expect(tracer.run(agent, {})).rejects.toThrowError("You need to buy credits to continue.");
  expect(process).toHaveBeenCalledTimes(2);
  expect(promptBuyCredits).toHaveBeenCalledTimes(2);
});

test("TerminalTracer should exit if user choose exit for buy credits", async () => {
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const tracer = new TerminalTracer(context);

  const select = spyOn(prompts, "select")
    .mockResolvedValueOnce("exit")
    .mockResolvedValueOnce("retry");
  expect(await tracer["promptBuyCredits"](new Error("You need to buy credits to continue."))).toBe(
    "exit",
  );
  expect(await tracer["promptBuyCredits"](new Error("You need to buy credits to continue."))).toBe(
    "retry",
  );

  select.mockRestore();
});

test("TerminalTracer should hold only one prompt at a time", async () => {
  const aigne = new AIGNE();
  const context = aigne.newContext();

  const tracer = new TerminalTracer(context);

  const start = Promise.withResolvers();

  const select = spyOn(prompts, "select").mockImplementationOnce((async () => {
    await start.promise;
    return "retry";
  }) as any);

  const results = Promise.all(
    new Array(10)
      .fill(0)
      .map(() => tracer["promptBuyCredits"](new Error("You need to buy credits to continue."))),
  );

  start.resolve();

  expect(await results).toEqual(new Array(10).fill("retry"));
  expect(select).toHaveBeenCalledTimes(1);

  select.mockRestore();
});
