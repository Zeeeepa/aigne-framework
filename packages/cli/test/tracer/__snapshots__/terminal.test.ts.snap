// Bun Snapshot v1, https://bun.sh/docs/test/snapshots

exports[`TerminalTracer should work correctly 1`] = `
{
  "message": "hello, this is a test response message",
}
`;

exports[`TerminalTracer should render output message with markdown highlight 1`] = `
"\x1B[90mâœ”\x1B[39m ðŸ¤– \x1B[90m(\x1B[39m\x1B[33m0\x1B[39m \x1B[90minput tokens\x1B[39m\x1B[32m, \x1B[39m\x1B[36m0\x1B[39m \x1B[90moutput tokens\x1B[39m\x1B[90m)\x1B[39m

\x1B[32m\x1B[1m## Hello\x1B[22m\x1B[39m

\x1B[0mI am from \x1B[34m\x1B[1mAIGNE\x1B[22m (\x1B[34m\x1B[4mhttps://www.aigne.io\x1B[24m\x1B[39m\x1B[34m)\x1B[39m\x1B[0m"
`;

exports[`TerminalTracer should render output message without markdown highlight in non-tty 1`] = `
"\x1B[90mâœ”\x1B[39m ðŸ¤– \x1B[90m(\x1B[39m\x1B[33m0\x1B[39m \x1B[90minput tokens\x1B[39m\x1B[32m, \x1B[39m\x1B[36m0\x1B[39m \x1B[90moutput tokens\x1B[39m\x1B[90m)\x1B[39m

## Hello
I am from [**AIGNE**](https://www.aigne.io)"
`;

exports[`TerminalTracer.marked should stripe code block meta 1`] = `
"\x1B[0mhello\x1B[0m

    \x1B[33m\x1B[34mfunction\x1B[39m\x1B[33m test() \x1B[39m{
    }

"
`;
